### PR TITLE
WebXRManager: Add support for WebXR layers.

### PR DESCRIPTION
**Pull request by cabanier**
This PR will add support for mono quad and cylinder WebXR Layers.
If the browser doesn't support these, they will be emulated. These layers will also be emulated in 2D mode.

Support for moving the layers or changing their layout can be added later.

The individual layer has its own render call and you can draw a scene to the layer contents as usual.

This contribution is funded by [Meta](https://meta.com)

### CODE REVIEW

**Comment by cabanier** [comment]
@mrdoob @Mugen87 what do you think of this PR?
Should I create an example with some UI? I can also do that later when I add the ability to move/change the layers/

**Comment by Mugen87** [comment]
> Should I create an example with some UI? I can also do that later when I add the ability to move/change the layers/

That sounds good to me. I think the existing example is sufficient for the start.

File: src/renderers/webxr/WebXRManager.js [hunk 1]
```diff
@@ -609,12 +687,224 @@ class WebXRManager extends EventDispatcher {
...
+
+			}
+
+			return plane.uuid;
+
+		};
+
+		this.createCylinderLayer = function ( radius, centralAngle, aspectratio, translation, quaternion, pixelwidth, pixelheight, rendercall ) {
+
+			const geometry = new CylinderGeometry( radius, radius, radius * centralAngle / aspectratio, 64, 64, true, Math.PI - centralAngle / 2, centralAngle );
```
Comment (davehill00) [inline src/renderers/webxr/WebXRManager.js:778]: Height segments can (and probably should) be 1, not 64.

**Comment by davehill00** [comment]
What do you think about creating the layer as an Object3D that can be placed in the scene, parented to other objects in the scene, etc.?

You're handling adding/removing the "hole punch" `plane` inside of WebXRManager, but what if you just let the user manage this part (possibly with a fallback of removing it automatically if a layer is destroyed, as you do now with the `plane`)?

I've been playing around with a slightly different integration, and have a `WebXRQuadLayer` object that looks like this:

`
class WebXRQuadLayer extends Object3D
{
	constructor( quadLayer, quadRenderTarget, glBinding ) {
		super();

		this.layer = quadLayer;
		this.renderTarget = quadRenderTarget;
		this.glBinding = glBinding;

		// create the hole-punch mesh matching the quad layer
		this.holePunchMesh = new Mesh(
			new PlaneGeometry(2.0 * quadLayer.width, 2.0 * quadLayer.height),
			new MeshBasicMaterial( { colorWrite: false } ) );
			
		// parent the hole-punch mesh to this object so that they move together in the scene
		this.add(this.holePunchMesh);
	}

	updateMatrixWorld( force ) {

		super.updateMatrixWorld( force );

		// Update the position of the quad layer to match the updated world matrix
		// @TODO - check force and/or matrixNeedsUpdate flags?
		// @TODO - I wish there was a way to not have to allocate a new object each time we do this.
		this.layer.transform = new XRRigidTransform(this.position, this.quaternion);

	}

	updateWorldMatrix( updateParents, updateChildren ) {

		super.updateWorldMatrix( updateParents, updateChildren );
		
		// Update the position of the quad layer to match the updated world matrix
		this.layer.transform = new XRRigidTransform(this.position, this.quaternion);

	}
}
`

This way, when I create the layer object, I just add it to my main scene (parented to whatever I want, animating it however I want, etc.) and the hole-punch and the quad layer are always in the right spot.

**Comment by cabanier** [comment]
@davehill00 [Object3D](https://threejs.org/docs/#api/en/core/Object3D) has a bunch of methods that don't apply to WebXR Layers.
The only thing that applies is position and orientation. Changing the geometry of Quad and Cylinder layers is also specific to each layer so we need another method anyway.

We should to keep them as separate objects outside the normal scene. @Mugen87 , is there a way to add hidden objects to a scene? right now the layer punch out is added to the main scene but it would be better if that was hidden from the author.

**Comment by davehill00** [comment]
@cabanier - reading through Object3D, I feel like most of what's there would make sense for a layer. If I think of the layer as a "portal into another scene", most of the actions I can take on an Object3D make sense on that "portal". All the parenting and transformation stuff feels valuable, as does the ray casting and beforeRender/afterRender callbacks. 

Were there specific behaviors or methods in particular you felt would not applicable to a layer?

I think there are a few things I'd like to accomplish here:
1. Make it easy to position and animate a "portal" in the context of a scene. E.G., using a layer to render a wrist-mounted "screen" would be much easier if I could just parent it to the player's wrist. 
2. Make it possible for a user to customize the hole-punch. I'm worried that your proposed solution abstracts this way too much -- e.g., how would I use a circular hole-punch with a quad layer in your proposed approach? I think creating the default hole-punch mesh makes sense -- this is the 90% case, but I'd still like to be able to override this if I need to.

**Comment by Mugen87** [comment]
>  is there a way to add hidden objects to a scene? right now the layer punch out is added to the main scene but it would be better if that was hidden from the author.

Something like that is not possible right now. All 3D objects of a scene graph are equivalent and "visible" in the respective data structure (the instance of `THREE.Scene`).

**Comment by cabanier** [comment]
@davehill00 I incorporated your suggestion and updated the implementation and the sample.
[Here](https://arvr-mastodon.s3.amazonaws.com/com.oculus.vrshell-20230111-133822.mp4)'s an video of the sample with layer manipulation and UI.

**Comment by cabanier** [comment]
@Mugen87 @davehill00 I decided not to create a custom Mesh class that checks for invalid matrices or other values. Please let me know what you think of the new code.
Changes:
- return a mesh instead of a uuid
- mesh visibility or if it's part of the main scene determines if it draws
- enhanced the example so you can move the cylinder layer, hide the quad layer and added an interactive UI layer
- used a custom geometry on the quad layer

**Comment by cabanier** [comment]
@Mugen87 what do I need to do to get this merged?

**Comment by Mugen87** [comment]
It would be good if @mrdoob approves the PR since new `WebXRManager` API is introduced. I want to avoid the situation when native layer scaling was implemented. I've prematurely merged the related PR #24877 and didn't notice the design flaws until @mrdoob drew attention to them.

**Comment by cabanier** [comment]
> It would be good if @mrdoob approves the PR since new `WebXRManager` API is introduced. 

Thanks!

**Comment by cabanier** [comment]
@mrdoob what do you think?

**Comment by cabanier** [comment]
@mrdoob @Mugen87 any comments?

**Comment by Mugen87** [comment]
The builds need an update so executing the example via `githack` does not work. You will have to wait until the PR is merged and the builds are updated. Or download the branch on your computer and execute `npm start`.

**Comment by mrdoob** [comment]
Sorry for the delay...

Do you mind simplifying the PR to just one WebXRLayer type? It's very hard to review right now...

Once we find the right approach we can then add the other types.

**Comment by cabanier** [comment]
> Do you mind simplifying the PR to just one WebXRLayer type? It's very hard to review right now...

No problem. I'll just support Quad layers for now.

**Comment by cabanier** [comment]
@mrdoob done.
From the PR, it looks like [src/renderers/webgl/WebGLTextures.js](https://github.com/mrdoob/three.js/pull/25254/files#diff-6bdac4da85be606f09e2ab24b41c761b23ced0d9acaa720e2f2200d938a714e9) changed a lot but all I did was add one condition at the start. I suspect the extra indents is making it look as if there are many changes.

**Comment by cabanier** [comment]
@mrdoob thanks for all these fixes!

**Comment by mrdoob** [comment]
I'll continue refactoring this week 👍

**Comment by cabanier** [comment]
> I'll continue refactoring this week 👍

Thank you! Please let me know if I can help out in any way

**Comment by cabanier** [comment]
> I'll continue refactoring this week 👍

@mrdoob were you able to make progress?

**Comment by cabanier** [comment]
ping @mrdoob any progress?

**Comment by cabanier** [comment]
ping @mrdoob any progress?

**Comment by cabanier** [comment]
or @Mugen87 ?

**Comment by mrdoob** [comment]
I didn't have access to my Quest 2 for the last few weeks and decided to buy a Quest Pro (ouch!)

Should be able to look into this next week.

**Comment by cabanier** [comment]
> I didn't have access to my Quest 2 for the last few weeks and decided to buy a Quest Pro (ouch!)

Nice! Let me know what you think of it.
> Should be able to look into this next week.

Thanks

**Comment by cabanier** [comment]
> I didn't have access to my Quest 2 for the last few weeks and decided to buy a Quest Pro (ouch!)
> 
> Should be able to look into this next week.

any progress?  Do I ned to redo the pr now that there are merge conflicts?

**Comment by cabanier** [comment]
> I didn't have access to my Quest 2 for the last few weeks and decided to buy a Quest Pro (ouch!)
> 
> Should be able to look into this next week.

Do you think this can be integrated or do I have to make changes?

**Comment by cabanier** [comment]
> I didn't have access to my Quest 2 for the last few weeks and decided to buy a Quest Pro (ouch!)
> 
> Should be able to look into this next week.

@mrdoob Do you think this can be integrated or do I have to make changes?

**Comment by cabanier** [comment]
I'm not sure. I made this PR in January but @mrdoob still hasn't integrated it.

**Pull request closed without merging**